Add support for clang with OpenMP and other minor changes to oneapi build system

msimberg opened the pull request:
Follow-up from #42653.

- Adds a `requires` for compilers to the `intel-oneapi-mkl` package to error out earlier during concretization rather than building. The downside of this is that the oneapi package definition also has a check for supported compilers and these could become out of sync, but I think the usability benefit is worth it.
- Adds a fallback option to look for libomp with clang (which doesn't report the path of `libomp.so` in newer versions).
- Minor error/logging message changes.

msimberg committed:
Add support for clang in oneapi packages with OpenMP

msimberg committed:
Add fallback search for libomp in OneApi package with OpenMP threading

msimberg committed:
Add requires for the compiler when using threads=openmp in intel-oneapi-mkl

msimberg committed:
Cosmetic changes to messages in oneapi.py

msimberg commented:
Just for reference, I tested this PR in DLA-Future's CI where it was able to find libomp with clang 15: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/6187989743#L1301 (on https://github.com/eth-cscs/DLA-Future/pull/1098 with commit 03857c62f29dcf940f2fd65d869981ccaf5ee37d, i.e. https://github.com/spack/spack/pull/42717/commits/03857c62f29dcf940f2fd65d869981ccaf5ee37d).

rscohn2 commented on lib/spack/spack/build_systems/oneapi.py:
```diff
@@ -206,20 +206,34 @@ def openmp_libs(self):
         else:
-            raise_lib_error("MKL OMP requires one of %gcc, %oneapi,or %intel")
+            raise_lib_error(
+                "MKL with OpenMP threading requires one of %clang, %gcc, %oneapi, or %intel"
```
```suggestion
                "OneAPI package with OpenMP threading requires one of %clang, %gcc, %oneapi, or %intel"
```

rscohn2 commented on lib/spack/spack/build_systems/oneapi.py:
```diff
@@ -206,20 +206,34 @@ def openmp_libs(self):
 
         omp_libs = LibraryList(omp_lib_path)
-        tty.info(f"mkl requires openmp library: {omp_libs}")
+        tty.info(f"MKL requires OpenMP library: {omp_libs}")
```
```suggestion
        tty.info(f"OneAPI package requires OpenMP library: {omp_libs}")
```

rscohn2 reviewed (dismissed):
Thanks! I hope this addresses the last obstacle for people to move from intel-mkl to intel-oneapi-mkl. If you prefer I can make the wording changes in another PR with the cleanup of the error function.

msimberg committed:
Update error message in oneapi.py

Co-authored-by: Robert Cohn <rscohn2@gmail.com>

msimberg committed:
Update another error message in oneapi.py

Co-authored-by: Robert Cohn <rscohn2@gmail.com>

msimberg committed:
Inline helper error function in oneapi.py

msimberg committed:
Update one more error message in oneapi.py

msimberg committed:
Wrap long line in oneapi.py

msimberg commented:
Thanks for having a look @rscohn2! I changed the messages according to your suggestions and inlined `raise_lib_error`. I also changed `"Cannot locate OpenMP library` to `"OneAPI package cannot locate OpenMP library`.
